File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/CustomDrawCardViewDemo/CustomDrawCardViewDemoMain.pas
unit CustomDrawCardViewDemoMain;

{$I cxVer.inc}

interface

uses
  Variants, Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxControls, cxLookupGrid, cxLookupDBGrid, cxGridCustomTableView,
  cxGridTableView, cxGridDBTableView, cxGridLevel, cxGridCustomView, cxGrid,
  cxGridCardView, cxGridDBCardView, cxMaskEdit, StdCtrls, ExtCtrls,
  cxContainer, cxEdit, cxTextEdit, cxDropDownEdit, cxDBEdit, cxStyles,
  Menus, ActnList, ImgList, ComCtrls, cxLookAndFeels, cxGraphics,
  ToolWin, cxCustomData, cxFilter, cxData, DB, cxDBData,
  cxClasses, cxDataStorage, cxDBLookupComboBox, cxBlobEdit, DemoUtils,
  BaseForm, cxLookAndFeelPainters, cxGridCustomLayoutView;

type
  TCustomDrawCardViewDemoMainForm = class(TfmBaseForm)
    miOptions: TMenuItem;
    miSeparator4: TMenuItem;
    ilPics: TImageList;
    miBlue: TMenuItem;
    miGold: TMenuItem;
    miGreen: TMenuItem;
    miGrey: TMenuItem;
    pnPersonLines: TPanel;
    pnPersonLinesCaption: TPanel;
    tvPersonLine: TcxGridDBTableView;
    lvPersonLine: TcxGridLevel;
    cxgPersonLine: TcxGrid;
    pnPersons: TPanel;
    cxgPersons: TcxGrid;
    cvPersons: TcxGridDBCardView;
    lvPersons: TcxGridLevel;
    Splitter: TSplitter;
    pnPersonsCaption: TPanel;
    tvPersonLineNAME: TcxGridDBColumn;
    cvPersonsFIRSTNAME: TcxGridDBCardViewRow;
    cvPersonsSECONDNAME: TcxGridDBCardViewRow;
    cvPersonsGENDER: TcxGridDBCardViewRow;
    cvPersonsBIRTHNAME: TcxGridDBCardViewRow;
    cvPersonsDATEOFBIRTH: TcxGridDBCardViewRow;
    cvPersonsBIRTHCOUNTRY: TcxGridDBCardViewRow;
    cvPersonsLOCATIONOFBIRTH: TcxGridDBCardViewRow;
    cvPersonsBIOGRAPHY: TcxGridDBCardViewRow;
    cvPersonsNICKNAME: TcxGridDBCardViewRow;
    miCustomDrawStyles: TMenuItem;
    miGradient: TMenuItem;
    miBackgroundImage: TMenuItem;
    miCar: TMenuItem;
    miSky: TMenuItem;
    miLoadImage: TMenuItem;
    OpenDialog: TOpenDialog;
    miDefaultDrawing: TMenuItem;
    FontDialog: TFontDialog;
    miFont: TMenuItem;
    miDependOnDataDrawing: TMenuItem;
    miEgypt: TMenuItem;
    miTile: TMenuItem;
    miMyFace: TMenuItem;
    cxStyleRepository1: TcxStyleRepository;
    stBlueDark: TcxStyle;
    stGold: TcxStyle;
    stBlueLight: TcxStyle;
    stBlueBright: TcxStyle;
    stYellowLight: TcxStyle;
    stGreyLight: TcxStyle;
    stBlueSky: TcxStyle;
    stDefault: TcxStyle;
    procedure cvPersonsCustomDrawCell(Sender: TcxCustomGridTableView;
      ACanvas: TcxCanvas; AViewInfo: TcxGridTableDataCellViewInfo; var ADone: Boolean);
    procedure miBlueClick(Sender: TObject);
    procedure miGreenClick(Sender: TObject);
    procedure miGoldClick(Sender: TObject);
    procedure miGreyClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure miCarClick(Sender: TObject);
    procedure miSkyClick(Sender: TObject);
    procedure miLoadClick(Sender: TObject);
    procedure miDefaultDrawingClick(Sender: TObject);
    procedure miFontClick(Sender: TObject);
    procedure miDependOnDataDrawingClick(Sender: TObject);
    procedure miMyFaceClick(Sender: TObject);
    procedure miEgyptClick(Sender: TObject);
    procedure miTileClick(Sender: TObject);
  private
    FCarBitmap, FSkyBitmap: TBitMap;
    FColorScheme: TColorScheme;
    FCustomDrawingStyle: TCustomDrawingStyle;
    FEgyptBitmap, FMyFaceBitmap, FTileBitmap: TBitMap;
    FFont: TFont;
    FUserDefinedImage, FCurrentBitmap: TBitMap;
    procedure GridsStyles;
    procedure AssignCustomDrawProc;
    procedure SetBkImage(AMenuItem: TObject; ABitMap: TBitMap);
    procedure SetGradientColor(AMenuItem: TObject;
      AColorScheme: TColorScheme; ABackgroudStyle: TcxStyle);
    procedure UncheckMenuItems;
  protected
    procedure LookAndFeelChanged(Sender: TcxLookAndFeel; AChangedValues: TcxLookAndFeelValues); override;
  end;

var
  CustomDrawCardViewDemoMainForm: TCustomDrawCardViewDemoMainForm;

implementation

{$R *.dfm}

uses
  FilmsDemoData, AboutDemoForm;

procedure TCustomDrawCardViewDemoMainForm.FormCreate(Sender: TObject);
begin
  FColorScheme := csBlue;
  GridsStyles;

  FCustomDrawingStyle := cdsBkImage;

  FFont := TFont.Create;
  FFont.Assign(cxgPersons.Font);

  FCarBitmap := TBitMap.Create;
  LoadImageFromRes(FCarBitmap, 'CAR');

  FSkyBitmap := TBitMap.Create;
  LoadImageFromRes(FSkyBitmap, 'SKY');

  FEgyptBitmap := TBitmap.Create;
  LoadImageFromRes(FEgyptBitmap, 'EGYPT');

  FMyFaceBitmap := TBitmap.Create;
  LoadImageFromRes(FMyFaceBitmap, 'MYFACE');

  FTileBitmap := TBitmap.Create;
  LoadImageFromRes(FTileBitmap, 'TILE');

  FUserDefinedImage := TBitmap.Create;
  FCurrentBitmap := FTileBitmap;

  FilmsDemoDM.cdsFilmsPersons.MasterSource := FilmsDemoDM.dsPersonLines;
  FilmsDemoDM.cdsFilmsPersons.IndexFieldNames := 'PersonLineID';
  FilmsDemoDM.cdsFilmsPersons.MasterFields := 'ID';
end;

procedure TCustomDrawCardViewDemoMainForm.FormDestroy(Sender: TObject);
begin
  FCurrentBitmap := nil;
  FUserDefinedImage.Free;
  FEgyptBitmap.Free;
  FMyFaceBitmap.Free;
  FTileBitmap.Free;
  FCarBitmap.Free;
  FSkyBitmap.Free;
end;

procedure TCustomDrawCardViewDemoMainForm.cvPersonsCustomDrawCell(
  Sender: TcxCustomGridTableView; ACanvas: TcxCanvas;
  AViewInfo: TcxGridTableDataCellViewInfo; var ADone: Boolean);
var
  ARec: TRect;
  ATextToDraw: String;
  ABool: Boolean;
  AIsVertical: Boolean;
  procedure SetTextToDraw;
  begin
    if (AViewInfo is TcxGridCardRowDataViewInfo) then
    begin
      if (AViewInfo.Item.Index = cvPersonsGENDER.Index) then
      begin
        if VarAsType(AViewInfo.GridRecord.Values[cvPersonsGENDER.Index], varBoolean) then
          ATextToDraw := 'Male'
        else
          ATextToDraw := 'Female';
      end
      else
        ATextToDraw := AViewInfo.GridRecord.DisplayTexts[AViewInfo.Item.Index];
    end
    else
      ATextToDraw := VarAsType(AViewInfo.Item.Caption, varString);
  end;

begin
  ARec := AViewInfo.Bounds;
  ACanvas.Canvas.Font.Assign(FFont);
  SetTextToDraw;
  case FCustomDrawingStyle of
    cdsBkImage:
      ACanvas.FillRect(ARec, FCurrentBitmap);
    cdsGradient:
    begin
      AIsVertical := (FColorScheme = csGrey) or (FColorScheme = csGold);
      if AViewInfo is TcxGridCardRowDataViewInfo then
        DrawGradient(ACanvas.Canvas, ARec, ColorScheme[Integer(FColorScheme), 1], ColorScheme[Integer(FColorScheme), 0], 40, AIsVertical)
      else
        DrawGradient(ACanvas.Canvas, ARec, ColorScheme[Integer(FColorScheme), 0], ColorScheme[Integer(FColorScheme), 1], 40, AIsVertical);
    end;
    cdsDependOnData:
    begin
      ACanvas.Canvas.Brush.Style := bsSolid;
      ACanvas.Canvas.Brush.Color := clBlueLight;
      ACanvas.Canvas.FillRect(ARec);
      if VarAsType(AViewInfo.GridRecord.Values[cvPersonsGENDER.Index], varBoolean) then
        ACanvas.Canvas.Font.Color := clBlue
      else
        ACanvas.Canvas.Font.Color := clFuchsia;
      if (AViewInfo is TcxGridCardRowDataViewInfo) and (AViewInfo.Item.Index = cvPersonsGENDER.Index) then
      begin
        ABool := VarAsType(AViewInfo.GridRecord.Values[cvPersonsGENDER.Index], varBoolean);
        ilPics.Draw(ACanvas.Canvas, ARec.Left + 2, ARec.Top, Integer(ABool));
        ATextToDraw := '';
      end;
    end;
  end;

  SetBkMode(ACanvas.Canvas.Handle, TRANSPARENT);
  ACanvas.DrawText(ATextToDraw, TcxCustomTextEditViewInfo(AViewInfo.EditViewInfo).TextRect, 0);

  ADone := FCustomDrawingStyle <> cdsDefaultDrawing;
end;

procedure TCustomDrawCardViewDemoMainForm.GridsStyles;
begin
  cvPersons.OptionsView.SeparatorColor := ColorScheme[2, 2];
  cvPersons.Styles.Background := stBlueSky;
end;

procedure TCustomDrawCardViewDemoMainForm.AssignCustomDrawProc;
begin
  if not Assigned(cvPersons.OnCustomDrawCell) then
    cvPersons.OnCustomDrawCell := cvPersonsCustomDrawCell;
end;

procedure TCustomDrawCardViewDemoMainForm.SetGradientColor(
  AMenuItem: TObject; AColorScheme: TColorScheme; ABackgroudStyle: TcxStyle);
begin
  if (FCustomDrawingStyle <> cdsGradient) or (FColorScheme <> AColorScheme) then
  begin
    FColorScheme := AColorScheme;
    if FCustomDrawingStyle <> cdsGradient then
    begin
      FCustomDrawingStyle := cdsGradient;
      UncheckMenuItems;
    end;
    MenuItemSetChecked(AMenuItem, True);
    cvPersons.OptionsView.SeparatorColor := ColorScheme[Integer(AColorScheme), 2];
    cvPersons.Styles.Background := ABackgroudStyle;
    AssignCustomDrawProc;
    cvPersons.Painter.Invalidate;
  end;
end;

procedure TCustomDrawCardViewDemoMainForm.LookAndFeelChanged(
  Sender: TcxLookAndFeel; AChangedValues: TcxLookAndFeelValues);
begin
  inherited LookAndFeelChanged(Sender, AChangedValues);
  UpdateTableViewStyleSheet(tvPersonLine);
end;

procedure TCustomDrawCardViewDemoMainForm.UncheckMenuItems;
begin
  MenuItemCheckSubItemWithTag('miGradient', -1);
  MenuItemCheckSubItemWithTag('miBackgroundImage', -1);
  MenuItemSetChecked('miDependOnDataDrawing', False);
  MenuItemSetChecked('miDefaultDrawing', False);
end;

procedure TCustomDrawCardViewDemoMainForm.miGreenClick(Sender: TObject);
begin
  SetGradientColor(Sender, csGreen, cxStyle2);
end;

procedure TCustomDrawCardViewDemoMainForm.miGoldClick(Sender: TObject);
begin
  SetGradientColor(Sender, csGold, stYellowLight);
end;

procedure TCustomDrawCardViewDemoMainForm.miGreyClick(Sender: TObject);
begin
  SetGradientColor(Sender, csGrey, stGreyLight);
end;

procedure TCustomDrawCardViewDemoMainForm.miBlueClick(Sender: TObject);
begin
  SetGradientColor(Sender, csBlue, stBlueSky);
end;

procedure TCustomDrawCardViewDemoMainForm.miLoadClick(Sender: TObject);
begin
  if OpenDialog.Execute then
  begin
    if FCustomDrawingStyle <> cdsBkImage then
    begin
      UncheckMenuItems;
      GridsStyles;
      FCustomDrawingStyle := cdsBkImage;
    end;
    MenuItemSetChecked(Sender, True);
    FUserDefinedImage.LoadFromFile(OpenDialog.FileName);
    FCurrentBitmap := FUserDefinedImage;
    AssignCustomDrawProc;
    cvPersons.Painter.Invalidate;
    SetCurrentDir(ExtractFilePath(Application.ExeName));
 end;
end;

procedure TCustomDrawCardViewDemoMainForm.miDefaultDrawingClick(Sender: TObject);
begin
  UncheckMenuItems;
  MenuItemSetChecked(Sender, True);
  FCustomDrawingStyle := cdsDefaultDrawing;
  cvPersons.OnCustomDrawCell := nil;
  cvPersons.Styles.Background := nil;
  cvPersons.OptionsView.SeparatorColor := clBtnFace;
  cvPersons.Painter.Invalidate;
end;

procedure TCustomDrawCardViewDemoMainForm.miFontClick(Sender: TObject);
begin
  FontDialog.Font.Assign(FFont);
  if FontDialog.Execute then
  begin
    FFont.Assign(FontDialog.Font);
    cvPersons.Styles.ContentEven.Font.Assign(FontDialog.Font);
    cvPersons.Styles.ContentOdd.Font.Assign(FontDialog.Font);
    cvPersons.LayoutChanged(False);
    cvPersons.Painter.Invalidate;
  end;
end;

procedure TCustomDrawCardViewDemoMainForm.miDependOnDataDrawingClick(Sender: TObject);
begin
  UncheckMenuItems;
  MenuItemSetChecked(Sender, True);
  FCustomDrawingStyle := cdsDependOnData;
  GridsStyles;
  AssignCustomDrawProc;
  cvPersons.Painter.Invalidate;
end;

procedure TCustomDrawCardViewDemoMainForm.miEgyptClick(Sender: TObject);
begin
  SetBkImage(Sender, FEgyptBitmap);
end;

procedure TCustomDrawCardViewDemoMainForm.miMyFaceClick(Sender: TObject);
begin
  SetBkImage(Sender, FMyFaceBitMap);
end;

procedure TCustomDrawCardViewDemoMainForm.miTileClick(Sender: TObject);
begin
  SetBkImage(Sender, FTileBitMap);
end;

procedure TCustomDrawCardViewDemoMainForm.miSkyClick(Sender: TObject);
begin
  SetBkImage(Sender, FSkyBitmap);
end;

procedure TCustomDrawCardViewDemoMainForm.miCarClick(Sender: TObject);
begin
  SetBkImage(Sender, FCarBitmap);
end;

procedure TCustomDrawCardViewDemoMainForm.SetBkImage(AMenuItem: TObject; ABitMap: TBitmap);
begin
  if FCustomDrawingStyle <> cdsBkImage then
  begin
    UncheckMenuItems;
    GridsStyles;
    FCustomDrawingStyle := cdsBkImage;
  end;
  MenuItemSetChecked(AMenuItem, True);
  FCurrentBitmap := ABitMap;
  AssignCustomDrawProc;
  cvPersons.Painter.Invalidate;
end;

end.  

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/CustomDrawCardViewDemo/CustomDrawCardViewDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "CustomDrawCardViewDemoMain.h"
#include "FilmsDemoData.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxGridCardView"
#pragma link "cxGridDBCardView"
#pragma link "cxLookAndFeels"
#pragma link "cxBlobEdit"
#pragma link "cxDataStorage"
#pragma link "cxDBLookupComboBox"
#pragma link "cxGridCustomLayoutView"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TCustomDrawCardViewDemoMainForm *CustomDrawCardViewDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TCustomDrawCardViewDemoMainForm::TCustomDrawCardViewDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------

String __fastcall TCustomDrawCardViewDemoMainForm::TextToDraw(TcxGridTableDataCellViewInfo *AViewInfo)
{
  if (dynamic_cast<TcxGridCardRowDataViewInfo*>(AViewInfo) != NULL) {
    if (AViewInfo->Item->Index == cvPersonsGENDER->Index) {
      if ((int)VarAsType(AViewInfo->GridRecord->Values[cvPersonsGENDER->Index], varBoolean))
        return("Male");
      else
        return("Female");
    }
    else
      return(AViewInfo->GridRecord->DisplayTexts[AViewInfo->Item->Index]);
  }
  else
    return(VarAsType(AViewInfo->Item->Caption, varString));
};
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::cvPersonsCustomDrawCell(
      TcxCustomGridTableView *Sender, TcxCanvas *ACanvas,
      TcxGridTableDataCellViewInfo *AViewInfo, bool &ADone)
{
  TRect ARec = AViewInfo->Bounds;
  ACanvas->Canvas->Font->Assign(FFont);

  String ATextToDraw = TextToDraw(AViewInfo);
  switch ((int)FCustomDrawingStyle) {
	case cdsBkImage: {
      ACanvas->FillRect(ARec, FCurrentBitmap);
    } break;
    case cdsGradient: {
      bool AIsVertical = ((FColorScheme == csGrey) || (FColorScheme == csGold));
      if (dynamic_cast<TcxGridCardRowDataViewInfo*>(AViewInfo) != NULL)
        DrawGradient(ACanvas->Canvas, ARec, ColorScheme[(int)FColorScheme][1], ColorScheme[(int)FColorScheme][0], 40, AIsVertical);
      else
        DrawGradient(ACanvas->Canvas, ARec, ColorScheme[(int)FColorScheme][0], ColorScheme[(int)FColorScheme][1], 40, AIsVertical);
    } break;
    case cdsDependOnData:
    {
      ACanvas->Canvas->Brush->Style = bsSolid;
      ACanvas->Canvas->Brush->Color = clBlueLight;
      ACanvas->Canvas->FillRect(ARec);
      if ((int)VarAsType(AViewInfo->GridRecord->Values[cvPersonsGENDER->Index], varBoolean))
        ACanvas->Canvas->Font->Color = clBlue;
      else
        ACanvas->Canvas->Font->Color = clFuchsia;
      if ((dynamic_cast<TcxGridCardRowDataViewInfo*>(AViewInfo) != NULL) && (AViewInfo->Item->Index == cvPersonsGENDER->Index)) {
        bool ABool = (int)VarAsType(AViewInfo->GridRecord->Values[cvPersonsGENDER->Index], varBoolean);
        ilPics->Draw(ACanvas->Canvas, ARec.Left + 2, ARec.Top, (int)ABool, true);
        ATextToDraw = "";
      } break;
    }
  }

  SetBkMode(ACanvas->Canvas->Handle, TRANSPARENT);
  ACanvas->DrawTexT(ATextToDraw, ((TcxCustomTextEditViewInfo*)(AViewInfo->EditViewInfo))->TextRect, 0, true);
  ADone = (FCustomDrawingStyle != cdsDefaultDrawing);
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::GridsStyles()
{
  	cvPersons->OptionsView->SeparatorColor = ColorScheme[2][2];
	cvPersons->Styles->Background = stBlueSky;
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::FormCreate(
      TObject *Sender)
{
  FColorScheme = csBlue;
  GridsStyles();

  FCustomDrawingStyle = cdsBkImage;

  FFont = new TFont();

  FCarBitmap = new Graphics::TBitmap();
  LoadImageFromRes(FCarBitmap, "CAR");

  FSkyBitmap = new Graphics::TBitmap();
  LoadImageFromRes(FSkyBitmap, "SKY");

  FEgyptBitmap = new Graphics::TBitmap();
  LoadImageFromRes(FEgyptBitmap, "EGYPT");

  FMyFaceBitmap = new Graphics::TBitmap();
  LoadImageFromRes(FMyFaceBitmap, "MYFACE");

  FTileBitmap =  new Graphics::TBitmap();
  LoadImageFromRes(FTileBitmap, "TILE");

  FUserDefinedImage = new Graphics::TBitmap();
  FCurrentBitmap = FTileBitmap;

  FilmsDemoDM->cdsFilmsPersons->MasterSource = FilmsDemoDM->dsPersonLines;
  FilmsDemoDM->cdsFilmsPersons->IndexFieldNames = "PersonLineID";
  FilmsDemoDM->cdsFilmsPersons->MasterFields = "ID";
}
//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::AssignCustomDrawProc()
{
  if (!cvPersons->OnCustomDrawCell)
    cvPersons->OnCustomDrawCell = cvPersonsCustomDrawCell;
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::UncheckMenuItem(TMenuItem* AMenuItem)
{
  for(int i=0; i < ComponentCount; i++)
	if((dynamic_cast<TMenuItem*>(Components[i]) != NULL) && (Components[i] != AMenuItem) &&
      ((TMenuItem*)Components[i])->Checked)
        ((TMenuItem*)Components[i])->Checked = false;
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::SetBkImage(TMenuItem* AMenuItem, Graphics::TBitmap* ABitMap)
{
  if(!((TMenuItem*)AMenuItem)->Checked) {
    ((TMenuItem*)AMenuItem)->Checked = true;
    if(FCustomDrawingStyle != cdsBkImage) {
      UncheckMenuItem(AMenuItem);
      GridsStyles();
      FCustomDrawingStyle = cdsBkImage;
      miBackgroundImage->Checked = true;
    }
    FCurrentBitmap = ABitMap;
    AssignCustomDrawProc();
    cvPersons->Painter->Invalidate();
  };
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::SetGradientColor(TMenuItem* AMenuItem,
  CustomDrawCardViewDemoTypesH::TColorScheme AColorScheme, TcxStyle *AStyle)
{
	if((FCustomDrawingStyle != cdsGradient) || (FColorScheme != AColorScheme))
	{
		FColorScheme = AColorScheme;
		AMenuItem->Checked = true;
		if (FCustomDrawingStyle != cdsGradient)
		{
			UncheckMenuItem(AMenuItem);
			FCustomDrawingStyle = cdsGradient;
			miGradient->Checked = true;
		};
		cvPersons->OptionsView->SeparatorColor = (TColor)(ColorScheme[(int)AColorScheme][2]);
		cvPersons->Styles->Background = AStyle;
		AssignCustomDrawProc();
    	cvPersons->Painter->Invalidate();
  	};
}

//---------------------------------------------------------------------------
void __fastcall TCustomDrawCardViewDemoMainForm::FormDestroy(
      TObject *Sender)
{
  FCurrentBitmap = NULL;
  delete FUserDefinedImage;
  delete FEgyptBitmap;
  delete FMyFaceBitmap;
  delete FTileBitmap;
  delete FCarBitmap;
  delete FSkyBitmap;
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miFontClick(
      TObject *Sender)
{
  FontDialog->Font->Assign(FFont);
  if (FontDialog->Execute()) {
    FFont->Assign(FontDialog->Font);
    cvPersons->Styles->ContentEven->Font->Assign(FontDialog->Font);
    cvPersons->Styles->ContentOdd->Font->Assign(FontDialog->Font);
    cvPersons->LayoutChanged(false);
    cvPersons->Painter->Invalidate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miTileClick(
      TObject *Sender)
{
  SetBkImage((TMenuItem*)Sender, FTileBitmap);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miSkyClick(
      TObject *Sender)
{
  SetBkImage((TMenuItem*)Sender, FSkyBitmap);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miEgyptClick(
      TObject *Sender)
{
  SetBkImage((TMenuItem*)Sender, FEgyptBitmap);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miMyFaceClick(
      TObject *Sender)
{
  SetBkImage((TMenuItem*)Sender, FMyFaceBitmap);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miCarClick(
      TObject *Sender)
{
  SetBkImage((TMenuItem*)Sender, FCarBitmap);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miLoadClick(TObject *Sender)
{
  if(OpenDialog->Execute()){
    ((TMenuItem*)Sender)->Checked = true;
    if(FCustomDrawingStyle != cdsBkImage){
      UncheckMenuItem((TMenuItem*)Sender);
      GridsStyles();
      FCustomDrawingStyle = cdsBkImage;
      miBackgroundImage->Checked = true;
    }

    FUserDefinedImage->LoadFromFile(OpenDialog->FileName);
    FCurrentBitmap = FUserDefinedImage;
    AssignCustomDrawProc();
    cvPersons->Painter->Invalidate();
    SetCurrentDir(ExtractFilePath(Application->ExeName));
  }
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miDependOnDataDrawingClick(
      TObject *Sender)
{
  if(!((TMenuItem*)Sender)->Checked) {
    ((TMenuItem*)Sender)->Checked = true;
    UncheckMenuItem((TMenuItem*)Sender);
    FCustomDrawingStyle = cdsDependOnData;
    GridsStyles();
    AssignCustomDrawProc();
    cvPersons->Painter->Invalidate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miDefaultDrawingClick(
      TObject *Sender)
{
  if (!((TMenuItem*)Sender)->Checked) {
    ((TMenuItem*)Sender)->Checked = true;
    UncheckMenuItem((TMenuItem*)Sender);
    FCustomDrawingStyle = cdsDefaultDrawing;
    cvPersons->OnCustomDrawCell = NULL;
    cvPersons->Styles->Background = NULL;
    cvPersons->OptionsView->SeparatorColor = clBtnFace;
    cvPersons->Painter->Invalidate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miGreyClick(TObject *Sender)
{
  SetGradientColor((TMenuItem*)Sender, csGrey, stGreyLight);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miGreenClick(TObject *Sender)
{
  SetGradientColor((TMenuItem*)Sender, csGreen, cxStyle2);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miGoldClick(TObject *Sender)
{
  SetGradientColor((TMenuItem*)Sender, csGold, stYellowLight);
}
//---------------------------------------------------------------------------

void __fastcall TCustomDrawCardViewDemoMainForm::miBlueClick(TObject *Sender)
{
  SetGradientColor((TMenuItem*)Sender, csBlue, stBlueSky);
}
//---------------------------------------------------------------------------
